starlite/asgi/routing_trie/traversal.py

```diff
@@ -106,16 +103,21 @@ def traverse_route_map(
             scope=scope,
         )
 
-    if has_path_param:
-        path_params.append(component)  # type: ignore[arg-type]
-
-        return traverse_route_map(
-            current_node=current_node["children"][PathParameterSentinel],
-            path=path,
-            path_components=path_components,
-            path_params=path_params,
-            scope=scope,
-        )
+    for type_ in current_node["child_path_parameter_types"]:
+        try:
+            type_(component)
+        except ValueError:
+            continue
+        else:
+            path_params.append(component)  # type: ignore[arg-type]
+
+            return traverse_route_map(
+                current_node=current_node["children"][type_],
+                path=path,
+                path_components=path_components,
+                path_params=path_params,
+                scope=scope,
+            )
```

Comment:
I'm thinking performance will be a thing here, but just wanted to get a POC going.